airbyte-integrations/bases/base-python/base_python/sdk/abstract_source.py

```diff
@@ -0,0 +1,148 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import copy
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Iterator, Mapping, MutableMapping, Tuple
+
+from airbyte_protocol import (
+    AirbyteCatalog,
+    AirbyteConnectionStatus,
+    AirbyteMessage,
+    AirbyteRecordMessage,
+    AirbyteStateMessage,
+    AirbyteStream,
+    ConfiguredAirbyteCatalog,
+    ConfiguredAirbyteStream,
+    Status,
+    SyncMode,
+)
+from airbyte_protocol import Type as MessageType
+
+from base_python.integration import Source
+from base_python.logger import AirbyteLogger
+from base_python.sdk.streams.core import Stream, IncrementalStream
+
+
+class AbstractSource(Source, ABC):
+    def __init__(self):
+        super().__init__()
+
+    @abstractmethod
+    def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> Tuple[bool, Any]:
+        """
+        :return: A tuple of (boolean, error). If boolean is true, then we can connect to the underlying data source using the provided configuration.
+        Otherwise, the input config cannot be used to connect to the underlying data source, and the "error" object should describe what went wrong.
+        The error object will be cast to string to display the problem to the user.
+        """
+
+    @abstractmethod
+    def streams(self, config: Mapping[str, Any] = None) -> Mapping[str, Stream]:
+        """
+        :return: A mapping from stream name to the stream class representing that stream
+        """
+
+    @property
+    def name(self) -> str:
+        """Source name"""
+        return self.__class__.__name__
+
+    def discover(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> AirbyteCatalog:
+        """Discover streams"""
+        streams = []
+
+        for name, stream in self.streams(config=config).items():
+            args = {
+                'name': name,
+                'json_schema': stream.get_json_schema(),
+            }
+
+            supported_sync_modes = [SyncMode.full_refresh]
+            if isinstance(stream, IncrementalStream):
```

Comment:
is it idiomatic in python to use instanceof for this? would it makes sense to expose a type enum to the stream interface and check that instead?

Reply:
merged incrementalStream and Stream so it is no longer relevant